bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/rewriter/gemini/QueryVisitorRewriter.java

```diff
@@ -0,0 +1,91 @@
+package com.google.zetasql.toolkit.antipattern.rewriter.gemini;
+
+import com.google.zetasql.SqlException;
+import com.google.zetasql.toolkit.antipattern.AntiPatternVisitor;
+import com.google.zetasql.toolkit.antipattern.cmd.InputQuery;
+import com.google.zetasql.toolkit.antipattern.rewriter.prompt.PromptYamlReader;
+import com.google.zetasql.toolkit.antipattern.util.AntiPatternHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+public class QueryVisitorRewriter {
+    private static final Logger logger = LoggerFactory.getLogger(QueryVisitorRewriter.class);
+    private final PromptYamlReader promptYamlReader;
+    private final AntiPatternHelper antiPatternHelper;
+
+    public QueryVisitorRewriter(AntiPatternHelper antiPatternHelper) throws IOException {
+        this.promptYamlReader = new PromptYamlReader();
+        this.antiPatternHelper = antiPatternHelper;
+    }
+
+    public String rewriteSQL(String inputQuery,
+                             AntiPatternVisitor visitor,
+                             Integer TTL) throws Exception {
+        TTL--;
+        if (TTL < 0) {
+            throw new Exception("LLM couldn't solve the specific anti pattern");
+        }
+        String prompt = this.promptYamlReader.getAntiPatternNameToPrompt().get(visitor.getName());
+
+        if (prompt == null) {
+            return  inputQuery; // No changes so it can use the same query
+        }
+
+        prompt = String.format(prompt, inputQuery).replace("%%","%");
+        String queryStr = GeminiRewriter.processPrompt(prompt, this.antiPatternHelper.getProject());
+
+        queryStr = checkAntiPattern(queryStr, visitor, TTL);
+
+        return queryStr;
+    }
+
+    public String checkAntiPattern(String queryStr,
+                                   AntiPatternVisitor visitor,
+                                   Integer TTL) throws Exception {
+        InputQuery inputQuery = new InputQuery(queryStr, "intermediate_rewrite");
+        List<AntiPatternVisitor> visitorsThatFoundAntiPatterns = new ArrayList<>();
+        try {
+            // parser visitors
+            this.antiPatternHelper.checkForAntiPatternsInQueryWithParserVisitors(inputQuery, visitorsThatFoundAntiPatterns);
```

Comment:
To my understanding this will check all 8 visitors. Should it not check just for the current visitor passed as an argument to this function?

My concern is that for each anti pattern found we will checkAntiPattern with every visitor available. Si if we have N input SQLs with M anti-patterns found each and a total of V visitors, we will have NxMxVxTTL visits ...

Reply:
Yep you are right, we should limit the visitors there by passing an argument